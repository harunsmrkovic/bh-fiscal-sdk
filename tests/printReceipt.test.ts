import { rest } from "msw";
import { setupServer } from "msw/node";
import FiscalSDK from "../src/index";

// Test instance
const fiscal = new FiscalSDK({
  host: "http://localhost:4000",
});

describe("Print receipt", () => {
  describe("when the service is working", () => {
    let lastBody: string;

    // Mocking server
    const server = setupServer(
      rest.post<string>(
        "http://localhost:4000/stampatifiskalniracun",
        async (req, res, ctx) => {
          return res(
            ctx.xml(`<?xml version="1.0" encoding="utf-8"?>
      <KasaOdgovor xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
        <Odgovori>
          <Odgovor>
            <Naziv>BrojFiskalnogRacuna</Naziv>
            <Vrijednost xsi:type="xsd:int">420</Vrijednost>
          </Odgovor>
          <Odgovor>
            <Naziv>DatumFiskalnogRacuna</Naziv>
            <Vrijednost xsi:type="xsd:string">01.01.2023.</Vrijednost>
          </Odgovor>
          <Odgovor>
            <Naziv>VrijemeFiskalnogRacuna</Naziv>
            <Vrijednost xsi:type="xsd:string">08:30:59</Vrijednost>
          </Odgovor>
          <Odgovor>
            <Naziv>IznosFiskalnogRacuna</Naziv>
            <Vrijednost xsi:type="xsd:double">10</Vrijednost>
          </Odgovor>
        </Odgovori>
        <VrstaOdgovora>OK</VrstaOdgovora>
        <BrojZahtjeva>233</BrojZahtjeva>
      </KasaOdgovor>`)
          );
        }
      )
    );

    beforeAll(() => server.listen());
    afterAll(() => server.close());

    beforeEach(() => {
      server.events.on("request:start", async (req) => {
        lastBody = await req.text();
      });
    });

    afterEach(() => {
      server.resetHandlers();
      lastBody = "";
      server.events.removeAllListeners("request:start");
    });

    it("should prepare and send the correct XML", async () => {
      await fiscal.printReceipt({
        articles: [
          {
            id: "1",
            name: "Zvake",
            price: 10,
            rate: "E",
            quantity: 1,
            discount: 0,
          },
        ],
        paymentMethods: [{ type: "Gotovina", amount: 10 }],
        billId: "1",
        date: new Date("2023-01-01T08:30:59"),
      });

      expect(lastBody).toMatchSnapshot();
    });

    it("should return the correct response", async () => {
      const response = await fiscal.printReceipt({
        articles: [
          {
            id: "1",
            name: "Zvake",
            price: 10,
            rate: "E",
            quantity: 1,
            discount: 0,
          },
        ],
        paymentMethods: [{ type: "Gotovina", amount: 10 }],
        billId: "1",
        date: new Date(),
      });

      expect(response).toStrictEqual({
        amount: 10,
        date: "01.01.2023.",
        id: 420,
        time: "08:30:59",
      });
    });
  });

  describe("when the service is not working", () => {
    // Mocking server
    const server = setupServer(
      rest.post<string>(
        "http://localhost:4000/stampatifiskalniracun",
        async (req, res, ctx) => {
          return res(
            ctx.xml(`<?xml version="1.0" encoding="utf-8"?>
      <KasaOdgovor xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
        <Odgovori>
          <Odgovor>
            <Naziv>Štampanje fiskalnog računa</Naziv>
            <Vrijednost xsi:type="xsd:int">ERROR_FISCAL_INVALID_ITEM_TAX</Vrijednost>
          </Odgovor>
        </Odgovori>
        <VrstaOdgovora>Greska</VrstaOdgovora>
        <BrojZahtjeva>233</BrojZahtjeva>
      </KasaOdgovor>`)
          );
        }
      )
    );

    beforeAll(() => server.listen());
    afterAll(() => server.close());
    afterEach(() => server.resetHandlers());

    it("should throw an error", async () => {
      await expect(
        fiscal.printReceipt({
          articles: [
            {
              id: "1",
              name: "Zvake",
              price: 10,
              rate: "E",
              quantity: 1,
              discount: 0,
            },
          ],
          paymentMethods: [{ type: "Gotovina", amount: 10 }],
          billId: "1",
          date: new Date(),
        })
      ).rejects.toThrowError("Error: ERROR_FISCAL_INVALID_ITEM_TAX");
    });
  });
});
