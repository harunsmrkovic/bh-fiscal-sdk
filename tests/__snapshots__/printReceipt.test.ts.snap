// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Print receipt when the service is working should prepare and send the correct XML 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<RacunZahtjev
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:xsd="http://www.w3.org/2001/XMLSchema">
	<BrojZahtjeva></BrojZahtjeva>
	<VrstaZahtjeva>0</VrstaZahtjeva>
	<NoviObjekat>
		<BrojRacuna>1</BrojRacuna>
		<Datum>2023-01-01T08:30:59</Datum>
		<StavkeRacuna>
			<RacunStavka>
				<artikal>
					<Sifra>1</Sifra>
					<Naziv>Zvake</Naziv>
					<JM></JM>
					<Cijena>10</Cijena>
					<Stopa>E</Stopa>
				</artikal>
				<Kolicina>1</Kolicina>
				<Rabat>0</Rabat>
			</RacunStavka>
		</StavkeRacuna>
		<VrstePlacanja>
			<VrstaPlacanja>
				<Oznaka>Gotovina</Oznaka>
				<Iznos>10</Iznos>
			</VrstaPlacanja>
		</VrstePlacanja>
	</NoviObjekat>
</RacunZahtjev>"
`;
